pkg/controllers/user/approuter/core_dns_client.go

```diff
@@ -0,0 +1,290 @@
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, errors.Wrap(err, "GetDomain: fail to get stored secret")
+	}
+
+	url := fmt.Sprintf("%s/domain/%s", c.base, fqdn)
+	req, err := c.request(http.MethodGet, url, nil)
+	if err != nil {
+		return d, errors.Wrap(err, "GetDomain: failed to build a request")
+	}
+
+	o, err := c.do(req)
+	if err != nil {
+		return d, errors.Wrap(err, "GetDomain: failed to execute a request")
+	}
+
+	return &o.Data, nil
+}
+
+func (c *Client) CreateDomain(hosts []string) (string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	url := fmt.Sprintf("%s/domain", c.base)
+	body, err := jsonBody(&model.DomainOptions{Hosts: hosts})
+	if err != nil {
+		return "", err
+	}
+
+	req, err := c.request(http.MethodPost, url, body)
+	if err != nil {
+		return "", errors.Wrap(err, "CreateDomain: failed to build a request")
+	}
+
+	resp, err := c.do(req)
+	if err != nil {
+		return "", errors.Wrap(err, "CreateDomain: failed to execute a request")
+	}
+
+	//to find token in management cluster namespace
+	if _, _, err = c.getSecret(); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return "", err
+		}
+		//if token not found, create a new one
+		if err = c.setSecret(&resp); err != nil {
+			return "", err
+		}
+	}
+
+	return resp.Data.Fqdn, err
+}
+
+func (c *Client) UpdateDomain(hosts []string) (string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	fqdn, token, err := c.getSecret()
+	if err != nil {
+		return "", errors.Wrap(err, "DeleteDomain: fail to get stored secret")
+	}
+
+	url := fmt.Sprintf("%s/domain/%s", c.base, fqdn)
+	body, err := jsonBody(&model.DomainOptions{Hosts: hosts})
+	if err != nil {
+		return "", err
+	}
+
+	req, err := c.request(http.MethodPut, url, body)
+	if err != nil {
+		return "", errors.Wrap(err, "UpdateDomain: failed to build a request")
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+
+	_, err = c.do(req)
+	if err != nil {
+		return "", errors.Wrap(err, "UpdateDomain: failed to execute a request")
+	}
+
+	return fqdn, nil
+}
+
+func (c *Client) DeleteDomain() (string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	fqdn, token, err := c.getSecret()
+	if err != nil {
+		return "", errors.Wrap(err, "DeleteDomain: fail to get stored secret")
+	}
+
+	url := fmt.Sprintf("%s/domain/%s", c.base, fqdn)
+	req, err := c.request(http.MethodDelete, url, nil)
+	if err != nil {
+		return "", errors.Wrap(err, "DeleteDomain: failed to build a request")
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+
+	_, err = c.do(req)
+	if err != nil {
+		return "", errors.Wrap(err, "DeleteDomain: failed to execute a request")
+	}
+
+	return fqdn, err
+}
+
+func (c *Client) RenewDomain() (string, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	fqdn, token, err := c.getSecret()
+	if err != nil {
+		return "", errors.Wrap(err, "RenewDomain: fail to get stored secret")
+	}
+
+	url := fmt.Sprintf("%s/domain/%s/renew", c.base, fqdn)
```

Comment:
we use `"%s/domain/%s` formatting in several places, can we extract it to the method?